### ISSUE TITLE
LED config question

### ISSUE THREAD

**Issue by giovanne123** [waldobronchart/AmbientLightServer#2]
For the LED config:
can I get it configured to only have one strang?
starting in the left bottom corner going not clockwise around, with 32 leds at bottom, 16 right, 32 top and 16 left ???

Is this correct?

In HardwareConfig.h I tried:
#define TOTAL_NUM_LEDS 96
#define NUM_LEDS_PER_STRAND 96
#define NUM_LEDS_HORIZONTAL 32
#define NUM_LEDS_VERTICAL 16

You pointed me also to: do I have to change something there?
https://github.com/waldobronchart/AmbientLightServer/blob/master/LEDController.cpp#L42

> To change how the LEDs are updated, you will need to make changes to LEDController::UpdateLEDs() function:
https://github.com/waldobronchart/AmbientLightServer/blob/master/LEDController.cpp#L42

What do you mean with how the LEDs are updated?

**Comment by waldobronchart** [comment]
You need to change the way the LEDs are updated, by changing the for loop in that function.

Line 62-84:
```
	// Update first strand of 25
	for (int i=0; i<NUM_LEDS_PER_STRAND; ++i)
	{
		Color color = lerpColor(m_prevColorBuffer[i], colorBuffer[i], lerpTerm);

		ShiftOut8Bits(GPIO_CLOCK_PIN1, GPIO_DATA_PIN1, color.ByteR());
		ShiftOut8Bits(GPIO_CLOCK_PIN1, GPIO_DATA_PIN1, color.ByteG());
		ShiftOut8Bits(GPIO_CLOCK_PIN1, GPIO_DATA_PIN1, color.ByteB());

		m_prevColorBuffer[i] = color;
	}

	// Update second strand of 25
	for (int i=NUM_LEDS_PER_STRAND; i<TOTAL_NUM_LEDS; ++i)
	{
		Color color = lerpColor(m_prevColorBuffer[i], colorBuffer[i], lerpTerm);

		ShiftOut8Bits(GPIO_CLOCK_PIN2, GPIO_DATA_PIN2, color.ByteR());
		ShiftOut8Bits(GPIO_CLOCK_PIN2, GPIO_DATA_PIN2, color.ByteG());
		ShiftOut8Bits(GPIO_CLOCK_PIN2, GPIO_DATA_PIN2, color.ByteB());

		m_prevColorBuffer[i] = color;
	}
```

If you have a single strand, remove the second for loop and change the first one so that it updates the right amount of LEDs.

**Comment by giovanne123** [comment]
Thanks, am I right
- I can remove the second for loop, ok
- but why do I have to change the first one, because in HardwareConfig.h there is already NUM_LEDS_PER_STRAND which is used in the first loop and I defined it to 96 in the config.

I will go one trying and testing :-) maybe I have to wait for the logitech or picam...

**Comment by waldobronchart** [comment]
Sure, you could keep the first loop as-is. Let me know if that works

**Comment by giovanne123** [comment]
One question about the pins where the LEDs strip (ws2801) can/should be connected? Because at the moment I can't get the LEDs light up corrosponding to the captured image. The LEDs are flickering in different colors and also not all the defined LEDs in HardwareConfig.h are used.

Therefore I came to the question (sorry about that, but I'm not so familiar with all this):
You defined the GPIOs for the LEDs in 0,1,2,3 wiringPi numbering
**I currently have the LED strip connected to 12 and 14 wiringPi numbering**, because this are the pins formally used from my hyperion installation. Hyperion used SPI (/dev/spidev0.0) and the pins 12 (MOSI) and 14 (CLK) are used for that.
**Are this two pins also usable for your Ambi-Solution?** (if not - hmm - I think I have to setup a testenvironment, because I don't want to change my hyperion inst. currently)  

About SPI, can this also an alternative to control the LEDs this way in your Ambi-Solution?

**Comment by waldobronchart** [comment]
I'm not familiar with Hyperion so I can't help you there.

For the pins, I can't remember off the top of my head. You should be looking at the pins layout for your Pi version and using that instead of the pins that I use.

**Comment by jimmyeao** [comment]
I can't figure out from your pictures which pins you have D1 and C1 connected to - and all the GPIO diagrams I see for the Pi don't have a pin 0, they start at 1. #confused

**Comment by waldobronchart** [comment]
These runtime errors seem to be related to opencv, not the pins. Make sure your webcam is working properly by capturing a picture from the command line.

**Comment by jimmyeao** [comment]
Hi,
I can capture images from the command line no problem, and when I run run.sh the led on the pi camera activates, and I can get an image from the ambilight client running on my windows server, so I suspect it the main it is working - any clue on the pins would be good :)

**Comment by waldobronchart** [comment]
A few things to check:

1. Check you are looking at the correct pins layout for your Raspberry Pi version. I used a version 1 Pi; you may have a different version.
2. If I recall correctly, you need to use the WiringPi layout instead of the Pi layout.

Hope that helps! It's been a long time since I did this project, my memory is a bit blurry on the details.

**Comment by jimmyeao** [comment]
Cool, thanks for your help, I have another stab at it later - using a Pi 3 for this project, planning to write it up from beginning to end includeding how to set up the ambilight client. I assume I am correct in thinking the ambilight client is only needed for setting up the server, it doesn’t need to run full time?

**Comment by jimmyeao** [comment]
@waldobronchart Thanks for the tip on the WiringPi, managed to identify the correct pin numbers now. Now I have to work out how to make the picam work with opencv, get lots of v42 errors in AmbilightServer at the beginning, the camera works fine from the command line..
I've created a fork as I'm using Opencv 3.3.0 and had to modify the makefile to include the correct libraries

**Comment by jimmyeao** [comment]
Awesome, now working with a usb camera, can not make it work with the pi camera. Next task is to calibrate the leds, yellows and greens are not coming through very well - any tips?

**Comment by waldobronchart** [comment]
I never got my LEDs completely 100% calibrated, they just don't have the same range as my TV did. It depends very much on which LEDs you are using. It's possible also that the webcam is distorting the colors because of low-light and therefore high ISO. Try using the client with fixed-color mode and testing out a few colors that way.

**Comment by jimmyeao** [comment]
Fixed color mode works fine, I think I'm struggling as I am only using one strand, it seems to be placing half the LEDs in the wrong place in relation to the screen, I may need to cut it in half and go to 2 strands to get the orientation correct. Almost perfected an install script with all the dependencies, just need to script it to start on boot and add some code to turn the LEDs off when a key is pressed. May look at wiring up an ir sensor so I can turn it on and off with my harmony remote. Runs really fast on a raspi 3. :)

**Comment by waldobronchart** [comment]
Notice [here](https://github.com/waldobronchart/AmbientLightServer/blob/master/LEDController.cpp#L42) that I'm updating two strands separately because linking them together didn't work well for me. There seemed to be a lot of noise at the end of the strands.

If your hardware configuration is different, you'll need to update this code too.

**Comment by jimmyeao** [comment]
Had a look at that code, decided I am going to go with the same setup as you, 2 strands - I haven't touched C++ for nearly 30 years :) working well with a Logitech C270 camera. This is probably my favorite project with the Raspberry Pi so far, thanks for doing it in the first place, and all the advice! I've forked the repo as I compiled it with a newer version of OpenCV which required some changes to the source code for additional includes (They seem to have split some of the functions out) and thought it may be useful to someone. Almost finished the install script to do all the pre reqs from a blank raspbian install - takes 3hours and 10 minutes on a Pi 3.
